lib/core/src/Cardano/Pool/Rank/Likelihood.hs

```diff
@@ -32,9 +32,10 @@ import Prelude
 
 import Cardano.Wallet.Primitive.Types
     ( ActiveSlotCoefficient (..)
-    , DecentralizationLevel (..)
+    , DecentralizationLevel
```

Comment:
Hi @HeinrichApfelmus!

I like that you've removed the default constructor for `DecentralizationLevel`.

I do wonder though whether we could improve this further by moving `DecentralizationLevel` into its own module: `Primitive.Types.DecentralizationLevel`.

`Primitive.Types` is large enough that it would be easy for someone (later on) to accidentally break the encapsulation boundary that you've created.

By moving `DecentralizationLevel` (and its associated functions) into its own module, it might make the encapsulation boundary clearer.

Reply:
While I agree in principle, adding a new module also has a cost (cognitive, potential recursive dependencies).

In my opinion, the `DecentralizationLevel` type is not interesting enough to merit its own module — actually, the entire type and its abstraction would be superfluous if we had stuck to the ledger convention for the *d* parameter. I don't want to give this type more attention than it deserves, making it an abstract data type just aims at curbing the worst abuse.

> I feel that Primitive.Types is large enough that it would be easy for someone (later on) to inadvertently break the encapsulation boundary. 

Alas, breaking the encapsulation boundary will still be as easy as exporting the constructor. 🤔 In general, I feel that encapsulation requires discipline that comes from the programmer, not discipline that is enforced by the compiler — the compiler can help, but it cannot replace.

Reply:
> actually, the entire type and its abstraction would be superfluous if we had stuck to the ledger convention for the d parameter. 

This is true.

If I recall, the motivation for inverting the value was to present users of the wallet with a percentage value that _increased_ as we made progress toward the goal of decentralization. So "100% decentralized" could be represented with a value of 100%. This value is also how it was displayed in Daedalus, IIRC.

(IIRC, the fact that internally the `d` value was supposed to _decrease_ from 1 to 0 initially caused a bit of confusion even among IOG engineers!)

Reply:
> So "100% decentralized" could be represented with a value of 100%. This value is also how it was displayed in Daedalus, IIRC.

I too believe that the idea of showing "100%" to the regular users is sound. 🤓 It's just that the conversion $d \to 1-d$ had drifted to the wrong API boundary (ledger/wallet as opposed to wallet/UX).